jgrapht-core/src/main/java/org/jgrapht/generate/DirectedScaleFreeGraphGenerator.java

```diff
@@ -0,0 +1,406 @@
+     *        If negative number, the user does not care about the total number of edges and is
+     *        interested only in the number of nodes, therefore, {@link #targetNodes} will be
+     *        considered instead. Otherwise, {@link #targetEdges} will be considered and
+     *        {@link #targetEdges} will be ignored.
+     * @param targetNodes Target number of nodes to reach. Zero is a valid value.<br>
+     *        This parameter has lower priority than {@link #targetEdges} and will be used only if
+     *        {@link #targetEdges} given is a <i>negative</i> number.
+     * @param rng The {@link Random} object to use.
+     */
+    public DirectedScaleFreeGraphGenerator(
+        float alpha, float gamma, float deltaIn, float deltaOut, int targetEdges, int targetNodes,
+        Random rng)
+    {
+        this.alpha = alpha;
+        this.alphaPlusBeta = 1.0f - gamma;
+        this.deltaIn = deltaIn;
+        this.deltaOut = deltaOut;
+        this.targetEdges = targetEdges;
+        this.targetNodes = targetNodes;
+        this.rng = Objects.requireNonNull(rng, "Random number generator cannot be null");
+
+        // Do several checks on the parameters
+        if (alpha < 0 || gamma < 0 || alpha + gamma > 1) {
+            throw new IllegalArgumentException(
+                String.format("alpha and gamma values of (%f, %f) are invalid", alpha, gamma));
+        }
+        if (deltaIn < 0 || deltaOut < 0) {
+            throw new IllegalArgumentException(
+                String
+                    .format(
+                        "deltaIn and deltaOut values of (%f, %f) are invalid", deltaIn, deltaOut));
+        }
+        if (targetEdges < 0 && targetNodes < 0) {
+            throw new IllegalArgumentException(
+                "can not have both targetEdges and targetNodes not set.");
+        }
+
+    }
+
+    /**
+     * Generates an instance of the {@link Graph}.
+     * 
+     * @param target the target graph
+     * @param resultMap not used by this generator, can be null
+     * 
+     * @throws TooManyFailuresException When the method fails {@link #maxFailures} times to add a
+     *         new edge to the growing graph.
+     */
+    @Override
+    public void generateGraph(Graph<V, E> target, Map<String, V> resultMap)
+    {
+        Set<V> newNodesSet = new HashSet<>();
+        Set<E> newEdgesSet = new HashSet<>();
+
+        if (targetEdges == 0 || (targetEdges < 0 && targetNodes == 0))
+            return;
+        V initV = target.addVertex();
+        newNodesSet.add(initV);
+
+        int failuresCounter = 0;
+        // grow network now, edge by edge. If the number of edges is unlocked, continue growing not
+        // only until targetNodes is reached, but until adding any more edges would add another
+        // node. i.e. allow more beta growth but not alpha nor gamma.
+        while (targetEdges >= 0 ? targetEdges > newEdgesSet.size()
+            : targetNodes >= newNodesSet.size())
```

Comment:
Same comment. Why is this `targetNodes >= newNodesSet.size()` and not simply `>`

Reply:
I already answered in my previous comment (on your first review).
To allow the graph to grow in **edges** _after the target number of **nodes** is reached_, as long as that will add no more nodes.
This is particularly important with high beta values, where the probability of adding new **internal** edges is higher than the probability of adding edges to/from new nodes.

Reply:
Somehow I never saw your previous answer. This makes sense, thanks.